drivers/spi/spi_ll_stm32.c

```diff
@@ -250,63 +250,135 @@ static int spi_stm32_get_err(SPI_TypeDef *spi)
 }
 
 /* Shift a SPI frame as master. */
-static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
+static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data, bool half_duplex)
 {
-	uint16_t tx_frame = SPI_STM32_TX_NOP;
-	uint16_t rx_frame;
-
-	while (!ll_func_tx_is_empty(spi)) {
-		/* NOP */
-	}
-
+	if (half_duplex) {
 #if defined(CONFIG_SOC_SERIES_STM32MP1X) || \
-	defined(CONFIG_SOC_SERIES_STM32H7X) || \
+    defined(CONFIG_SOC_SERIES_STM32H7X) || \
 	defined(CONFIG_SOC_SERIES_STM32U5X)
 	/* With the STM32MP1, STM32U5 and the STM32H7,
 	 * if the device is the SPI master,
 	 * we need to enable the start of the transfer with
 	 * LL_SPI_StartMasterTransfer(spi)
 	 */
-	if (LL_SPI_GetMode(spi) == LL_SPI_MODE_MASTER) {
-		LL_SPI_StartMasterTransfer(spi);
-		while (!LL_SPI_IsActiveMasterTransfer(spi)) {
-			/* NOP */
+		if (LL_SPI_GetMode(spi) == LL_SPI_MODE_MASTER) {
+			LL_SPI_StartMasterTransfer(spi);
+			while (!LL_SPI_IsActiveMasterTransfer(spi)) {
+				/* NOP */
+			}
 		}
-	}
 #endif
 
-	if (SPI_WORD_SIZE_GET(data->ctx.config->operation) == 8) {
-		if (spi_context_tx_buf_on(&data->ctx)) {
-			tx_frame = UNALIGNED_GET((uint8_t *)(data->ctx.tx_buf));
+		if (SPI_WORD_SIZE_GET(data->ctx.config->operation) == 8) {
+			if (spi_context_tx_buf_on(&data->ctx)) {
+				uint8_t tx_frame = UNALIGNED_GET((uint8_t *)(data->ctx.tx_buf));
+				LL_SPI_TransmitData8(spi, tx_frame);
+				while (!ll_func_tx_is_empty(spi)) {
+					/* NOP */
+				}
+				/* Wait for transmit to finish */
+				while (ll_func_spi_is_busy(spi)) {
+					/* NOP */
+				}
+			}
+			spi_context_update_tx(&data->ctx, 1, 1);
```

Comment:
I think spi_context_update_tx need put to previous line.
because when spi transmit sucessed  then update spi context.

Reply:
And I think there is no need split half_duplex code and funll_fuplex code.
only add ` if ( half_duplex ) return;` after spi_context_update_tx.